Logs API to have functionality for reusing Standard Attributes (#4373)
open-telemetry/opentelemetry-specification: specification/logs/api.md

- [Severity Number](./data-model.md#field-severitynumber) (optional)
- [Severity Text](./data-model.md#field-severitytext) (optional)
- [Body](./data-model.md#field-body) (optional)
- [Attributes](./data-model.md#field-attributes) (optional) [1]
- **Status**: [Development](../document-status.md) - [Event Name](./data-model.md#field-eventname) (optional)

> [1]: **Status**: [Development](../document-status.md) - The API SHOULD provide ergonomics so that the caller can reuse

Reviewer: not blocking this PR:  I consider this as a bugfix and wonder if we need this note to be Development. I believe most SDKs already implement it this way (while the extended attribute support is very limited).

Let's bring it up on the spec call tomorrow?

Author: Personally, I see it like an minor enhancement.
However, maybe @open-telemetry/technical-committee will find it acceptable to make it as part of the stable spec as it should not be distributing.

Author: During Spec SIG meeting we agreed that it would be safer the have a seperate PR that removes the `Development` status. The PR description will describe the reasons why we are stabilizing it (so soon).

> [Standard Attributes](../common/README.md#standard-attribute)
> that are used when emitting telemetry for other signals.

Reviewer: may be worth mentioning extended attributes (since many SDKs don't support them on log records)?

E.g.
```suggestion
[Standard Attributes](../common/README.md#standard-attribute)
 along with extended attributes allowing values of [type Any](/specification/logs/data-model.md#type-any).
```

Author: > since many SDKs don't support them on log records

I think it is redundant as the place where the footnote/reference is used has a reference to this type.

Author: PTAL https://github.com/open-telemetry/opentelemetry-specification/pull/4373/commits/c4f644a7374e5ea869036d5496af838658dcaf65

- [Severity Number](./data-model.md#field-severitynumber) (optional)
- [Severity Text](./data-model.md#field-severitytext) (optional)
- [Body](./data-model.md#field-body) (optional)
- [Attributes](./data-model.md#field-attributes) (optional) [1]
- **Status**: [Development](../document-status.md) - [Event Name](./data-model.md#field-eventname) (optional)

> [1]: **Status**: [Development](../document-status.md) -

Reviewer 2: nit: Unless the person reading this is already aware that Attributes for Log is a superset of other attributes, this may not be obvious.
We do clarify that in https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-attributes , so wondering if this new change is also nearby that section, or atleast link to it, so readers immediately know the attributes here is different and hence the need for ergonomic conversion..

Author: https://github.com/open-telemetry/opentelemetry-specification/pull/4373/commits/dff1617c8e5cbe25967ff6b426bff881a7be4f0e

> The API SHOULD provide ergonomics so that the caller can use

Reviewer 3: Worth updating the spec compliance matrix? https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md#logs

Author: Done https://github.com/open-telemetry/opentelemetry-specification/pull/4373/commits/fe89a786d42bcd50bc45e23aa762df57f7fe7218

[Standard Attributes](../common/README.md#standard-attribute)
along with log attributes of [type `map<string, any>`](./data-model.md#type-mapstring-any).

> This allows the reuse of attributes across signals.

Reviewer 2: > This allows the reuse of attributes across signals.

is the ergonomic conversion one-way? i.e standard attributes are be used in places expecting LogAttributes, but not vice-versa?

Reviewer 2: If one-way, then the following maybe a better phrasing:
_This allows the reuse of attributes from Traces, Metrics to Logs as well._

Author: https://github.com/open-telemetry/opentelemetry-specification/pull/4373/commits/dff1617c8e5cbe25967ff6b426bff881a7be4f0e

> This allows the reuse of attributes when emitting telemetry for other signals.

Reviewer 4: ```suggestion
This allows the reuse of attributes across signals.
```

- [Severity Number](./data-model.md#field-severitynumber) (optional)
- [Severity Text](./data-model.md#field-severitytext) (optional)
- [Body](./data-model.md#field-body) (optional)
- [Attributes](./data-model.md#field-attributes) (optional) [1]
- **Status**: [Development](../document-status.md) - [Event Name](./data-model.md#field-eventname) (optional)

> [1]: **Status**: [Development](../document-status.md) -
> The API SHOULD provide ergonomics so that the caller can use

Reviewer 5: Having a normative recommendation in a note makes this easy to miss. We should make it clear in its own paragraph.

```suggestion
[1]: **Status**: [Development](../document-status.md)

The API SHOULD provide ergonomics so that the caller can use
```

Author: IMO this does not render nicely :cry:

Author: I think that the footnote should be in one paragraph. But maybe just better to make it a separate paragraph instead of a footnote? (without `[1]: `)

Author: PTAL https://github.com/open-telemetry/opentelemetry-specification/pull/4373/commits/a91104dca74fb5f7db388832eb7191cb42033007

> The API SHOULD provide ergonomics so that the caller can use
> [Standard Attributes](../common/README.md#standard-attribute)
> along with [Log Attributes](./data-model.md#field-attributes)
> of [type `map<string, any>`](./data-model.md#type-mapstring-any).
> This allows the reuse of [Standard Attributes](../common/README.md#standard-attribute)
> across signals.
> This can be converting functions, method overloads, etc.

Reviewer 5: Ergonomics is the study of people's efficiency in their working environment. Can we be more explicit about what is being recommended here?

```suggestion
The API SHOULD provide functionality for users to convert from
[Standard Attributes](../common/README.md#standard-attribute)
to [Log Attributes](./data-model.md#field-attributes)
of [type `map<string, any>`](./data-model.md#type-mapstring-any).
This allows the reuse of [Standard Attributes](../common/README.md#standard-attribute)
across signals.
```

Author: What do you think of

```suggestion
The API SHOULD provide functionality so that the caller can use
[Standard Attributes](../common/README.md#standard-attribute)
along with [Log Attributes](./data-model.md#field-attributes)
of [type `map<string, any>`](./data-model.md#type-mapstring-any).
This allows the reuse of [Standard Attributes](../common/README.md#standard-attribute)
across signals.
This can be converting functions, method overloads, etc.
```

Reviewer 5: The last sentence is redundant. "Functionality" is used throughout the specification to imply what that sentence is saying.

Also, we do not plan for users to use standard attributes along with log attributes. We plan for them to convert standard attributes into log attributes and use those.

Author: > we do not plan for users to use standard attributes along with log attributes. 

I think that some languages may want to do this.

Author: They want to have a `LogRecordBuilder` that have one method that accepts Standard Attributes and another one that accepts Log Attributes. It is like of we would add `AddStandardAttributes` to `Record` in Go Logs API.

Reviewer 5: Do they not plan to provide functionality to convert standard to log attributes?

I do not want to add functionality to the Go log signal to allow standard attributes to be used along with log attributes. I think [this](https://github.com/open-telemetry/opentelemetry-go/pull/6180) approach is better.

Author: https://github.com/open-telemetry/opentelemetry-specification/pull/4373/commits/28fb22afaadec5dc5192943aadeb05fe1024b7a3
